File: bilan_sky/bilan_air_booking_system/utils/rich_text.py
"""Convert Desk Text Editor (Quill) HTML to plain text for portal text fields."""

from __future__ import annotations

import re

from frappe.utils import strip_html


def rich_text_to_plain(text: str | None) -> str:
	"""Strip HTML from Text Editor fields; preserve basic line breaks."""
	if not text:
		return ""
	html = str(text)
	html = re.sub(r"<br\s*/?>", "\n", html, flags=re.IGNORECASE)
	html = re.sub(r"</(?:p|div|li|h[1-6]|tr)[^>]*>", "\n", html, flags=re.IGNORECASE)
	plain = strip_html(html)
	plain = re.sub(r"[ \t]+\n", "\n", plain)
	plain = re.sub(r"\n{3,}", "\n\n", plain)
	return plain.strip()

File: frontend/lib/rich-text.ts
/** Strip Desk Text Editor (Quill) HTML for plain text fields. */
export function richTextToPlain(html: string | null | undefined): string {
  if (!html) return "";
  let text = html
    .replace(/<br\s*\/?>/gi, "\n")
    .replace(/<\/(?:p|div|li|h[1-6]|tr)[^>]*>/gi, "\n")
    .replace(/<[^>]+>/g, "")
    .replace(/&nbsp;/gi, " ")
    .replace(/&amp;/gi, "&")
    .replace(/&lt;/gi, "<")
    .replace(/&gt;/gi, ">")
    .replace(/&quot;/gi, '"')
    .replace(/&#39;/gi, "'");
  text = text.replace(/[ \t]+\n/g, "\n").replace(/\n{3,}/g, "\n\n");
  return text.trim();
}

/** One-line preview for list rows. */
export function richTextPreview(html: string | null | undefined, maxLen = 80): string {
  const text = richTextToPlain(html).replace(/\s+/g, " ").trim();
  if (!text) return "—";
  return text.length > maxLen ? `${text.slice(0, maxLen)}…` : text;
}
